cluster/eval_server/add_model.py

```diff
@@ -0,0 +1,155 @@
+LIB_TF_CC_SO = "cc/tensorflow/lib/libtensorflow_cc.so"
+
+
+def verify_params(model_path, flags, binary_path):
+    # Verify model_path = .../<model>.pb
+    assert model_path.endswith('.pb'), model_path
+
+    assert flags, "Flags must be non-empty"
+    assert "-num_readouts=" in flags, "must set playouts"
+    assert "-model" not in flags, "leave model unset"
+
+    # Verify binary_path is to a gtp binary and it exists
+    binary_base = binary_path.replace(GTP_BIN, "")
+    assert binary_path.endswith(GTP_BIN), binary_path
+    assert os.path.isfile(binary_path)
+    # $ ldd bazel-bin/cc/gtp
+    #    libtensorflow_framework.so => (bazel path) => cc/tensorflow/lib/libtensorflow_framework.so
+	#    libtensorflow_cc.so => (bazel path) => cc/tensorflow/lib/libtensorflow_cc.so
+    #    libcuda.so.1
+    #    libnvidia-fatbinaryloader.so.<version e.g. 390>.116
+    assert os.path.isfile(os.path.join(binary_base, LIB_TF_FW_SO))
+    assert os.path.isfile(os.path.join(binary_base, LIB_TF_CC_SO))
+
+    return True
+
+def copy_to_gcs(src, dst):
+    assert gfile.Exists(src)
+    assert not gfile.Exists(dst)
+
+    with gfile.GFile(src, "rb") as src_f, gfile.GFile(dst, "wb") as dst_f:
+        shutil.copyfileobj(src_f, dst_f)
+
+
+def add_model(name, model_path, flags, binary_path):
+    """Add a player (model + flags) to models cbt
+    name: name of the model
+    model_path: path to model
+    flags: flags used to play this model
+    binary_path: path to gtp binary that can execute model
+    """
+
+    assert verify_params(model_path, flags, binary_path)
+
+    now_date = datetime.datetime.now().isoformat(' ')
+
+    metadata = (
+        (MODEL_NAME, name),
+        (b"model_num", ""),
+        (b"model_flags", flags),
+        (b"run", ""),
+        (b"parent", ""),
+        (b"tag", "eval_server"),
+        (b"tool", "add_model.py"),
+        (b"trained_date", now_date),
+    )
+
+    hash_params = model_path.encode()
+    hash_params += flags.replace(' ', '').replace('=', '').encode()
+    # ten matches MODEL_ROW format
+    player_hash = hashlib.md5(hash_params).hexdigest()[:10]
+    player_name = MODEL_ROW.format("z_eval", player_hash)
+
+    # Check that model doesn't already uploaded.
+    bt_table = (bigtable
+                .Client(FLAGS.cbt_project, admin=True)
+                .instance(FLAGS.cbt_instance)
+                .table(FLAGS.cbt_table))
+    assert bt_table.exists(), f"Table({FLAGS.cbt_table}) doesn't exist"
+    assert bt_table.read_row(player_name) is None, f"{player_name} already uploaded"
+
+    # Upload model and binary
+    gcs_player_folder = os.path.join(PLAYER_FOLDER, player_hash)
+    new_model_path = os.path.join(gcs_player_folder, os.path.basename(model_path))
+    new_binary_base = os.path.join(gcs_player_folder, "bin")
+    binary_base = binary_path.replace(GTP_BIN, "")
+
+    print (f"Saving to {gcs_player_folder!r}")
+
+    #copy_to_gcs(model_path, new_model_path)
```

Comment:
The print statement above is a lie?

(i'm assuming this is pending some clever way to dedupe the tf libs so as not to reup the same ones over and over?)

Reply:
fixed.